### PR TITLE
ESQL QSTR function

### PR DESCRIPTION
**Pull request by carlosdelest**
Adds a QSTR function, behind snapshot builds.

This is very much WIP for sharing ideas and early feedback.

Current approach:
- Use a `FullTextFunction` base class for full text query functions
- `QueryStringFunction` returns a boolean - the type was done just to fit as part of WHERE boolean expressions, conceptually it returns a query to be executed at Lucene. This may need its own data type 🤔 
- `FullTextFunction`s are able to be pushed down to Lucene as part of the local physical plan optimizing, similar to MATCH operator
- Similar to how the `MATCH` command worked, we're limiting the commands that can be used before the function, so we don't allow commands that may alter the existing columns (`DROP`, `KEEP`, `EVAL`).

### CODE REVIEW

**Comment by carlosdelest** [comment]
@elasticsearchmachine run buildkite/docs-build-pr

File: x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/function/fulltext/QueryStringFunction.java
Comment (fang-xing-esql) [inline x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/function/fulltext/QueryStringFunction.java:30]: Usually an ES|QL function has its own `AbstractFunctionTestCase`, an important usage is to generate documents(some of them are used by Kibana) automatically. This function doesn't have its compute engine implementation, I'm not quite sure how to make an `AbstractFunctionTestCase` for it yet, need to figure out how to generate the docs automatically, or manually create them.

Comment (fang-xing-esql) [reply x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/function/fulltext/QueryStringFunction.java:30]: `QueryStringFunction` takes only string literal, however if a numeric type is provided, it passes `Analyzer`, if a field is provided as input, it fails at ExpressionTranslator, these could be candidates for `VerifierTests`.

Override `resolveType()` is the usual way to validate input data types and foldable for functions during `Analyzer` phase, perhaps refer to an ES|QL string function as an example.

**Review by fang-xing-esql — COMMENTED** [review]
Thanks for putting these together @carlosdelest, my comments are added.

File: x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/function/fulltext/QueryStringFunction.java
Comment (ioanatia) [reply x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/function/fulltext/QueryStringFunction.java:30]: Good point on  `AbstractFunctionTestCase`, but we don't want generate documentation yet since the function is only available in snapshots. Also I might be looking at the wrong place, but we have another function `Rate` that is only available in snapshots and it also seems to be missing its own `AbstractFunctionTestCase`.
So I wonder if we can defer the need for a `AbstractFunctionTestCase` for now, since it seems it requires a bit of rework so that `AbstractFunctionTestCase` can also work with functions that do not have a compute engine implementation.

Comment (carlosdelest) [reply x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/function/fulltext/QueryStringFunction.java:29]: > QueryStringFunction takes only string literal, however if a numeric type is provided, it passes Analyzer, if a field is provided as input, it fails at ExpressionTranslator, these could be candidates for VerifierTests.
> 
> Override resolveType() is the usual way to validate input data types and foldable for functions during Analyzer phase, perhaps refer to an ES|QL string function as an example.

I've tried to address this in 7b2cbaa151cd60b6ba71a8aeb541463ef59ad344

Comment (carlosdelest) [reply x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/function/fulltext/QueryStringFunction.java:29]: ++ Ioana on documentation and `AbstractFunctionTestCase` - we can probably defer this until we plan to move it out of snapshot builds

File: x-pack/plugin/esql/src/test/java/org/elasticsearch/xpack/esql/optimizer/LocalPhysicalPlanOptimizerTests.java
Comment (carlosdelest) [reply x-pack/plugin/esql/src/test/java/org/elasticsearch/xpack/esql/optimizer/LocalPhysicalPlanOptimizerTests.java:578]: Thanks @fang-xing-esql ! I've added these tests in c5b45ec0f43cd7b621ba306be90431093217eb2c.

Are there any other combinations that you can think of, or any rule of thumb that I could use to understand what cases should I be implementing? I haven't found more examples in `LocalPhysicalPlanOptimizerTests` that could guide me here.

>  it looks like multiple qstrs are allowed but not combined for now, perhaps it can be a candidate for further optimization.

Probably so - I wonder if queries pushed down to Lucene should be left to Lucene to optimize? 🤔

Comment (fang-xing-esql) [reply x-pack/plugin/esql/src/test/java/org/elasticsearch/xpack/esql/optimizer/LocalPhysicalPlanOptimizerTests.java:578]: Thanks for adding more tests!

> Are there any other combinations that you can think of, or any rule of thumb that I could use to understand what cases should I be implementing? I haven't found more examples in `LocalPhysicalPlanOptimizerTests` that could guide me here.

It could be challenging to write effective testcases. We always try to improve test coverage, but it is not guaranteed that everything is covered, there could be many edge cases, I found PR reviews help me a lot to improve test coverage. There are many existing test cases, and it is normal not be able to remember all of them :). These are the things that come to my mind first when writing tests for new functions:

1. Verify resolutions - function and type resolution. This is done(majority of them) by `Analyzer`,  and these tests are usually covered by `AnalyzerTests` and `VerifierTests`.
2. Verify optimizer builds an optimal/desired plan. These are usually covered by `LogicalPlanOptimizerTests`, `PhysicalPlanOptimizerTests`, `QueryTranslatorTests`, and their local version if exist. For the functions and operators that can appear under where clause, I would consider test compound predicates - conjunctive and disjunctive. For this function in particular, because it got pushed down into Lucene, we could get some combinations of bool, should, must queries, which makes me think that we may want to cover conjunctive and disjunctive, and perhaps some nested cases. Most of the ES|QL functions have their own compute engine implementation, and this doesn't, which makes me think that we may want to cover some combinations of them. I think these are not edge cases, they might be quite common use cases. I'll try to think about more scenarios, and update it here, just don't want to throw all of them at one time :-).
3. Verify the query returns correct results. These are mainly done in CsvTests, REST and yml tests. I would add more complicated(conjunctive and disjunctive) tests to verify we get correct results.

> 
> > it looks like multiple qstrs are allowed but not combined for now, perhaps it can be a candidate for further optimization.
> 
> Probably so - I wonder if queries pushed down to Lucene should be left to Lucene to optimize? 🤔

Yeah, Lucene might be able to do optimizations on multiple queries pushed down to it. Honestly I'm not an expert of it, still learning. There is something we can do in optimizer, an example of it is `CombineBinaryComparisons`, if Lucene doesn't do it or doesn't do it well, this rule can be a reference.

File: x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/function/fulltext/QueryStringFunction.java
Comment (fang-xing-esql) [reply x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/function/fulltext/QueryStringFunction.java:29]: I added a ui label, so that Kibana is aware of this. As far as I know `AbstractFunctionTestCase` generates docs for Kibana, we can figure out the best option to move forward.

File: x-pack/plugin/esql/qa/testFixtures/src/main/resources/qstr-function.csv-spec
Comment (fang-xing-esql) [inline x-pack/plugin/esql/qa/testFixtures/src/main/resources/qstr-function.csv-spec:2]: I would recommend to add a few more complicated tests here, some combinations of conjunctive and disjunctive predicates, qrst with other functions - functions that can be pushed down into Lucene and also have compute engine implementation(like cider_match), functions that cannot be pushed down into Lucene and but have compute engine implementation(like length), perhaps pick/duplicate a few from the new tests added in LocalPhysicalPlanOptimizerTests and added here, as this is the place we validate correct results.

File: x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/analysis/Verifier.java
Comment (fang-xing-esql) [inline x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/analysis/Verifier.java:421]: Can this be moved into `checkFullTextQueryFunctions` so that there is one place to validate FullTextQuery related restrictions?

File: x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/analysis/Verifier.java
Comment (fang-xing-esql) [inline x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/analysis/Verifier.java:686]: The normal/usual way to validate foldable and input data types are in the `resolveType()` method of the function itself, to make it more manageable. Please refer to one of the ES|QL string functions, e.g. `Concat`.

File: x-pack/plugin/esql/src/test/java/org/elasticsearch/xpack/esql/analysis/VerifierTests.java
Comment (fang-xing-esql) [inline x-pack/plugin/esql/src/test/java/org/elasticsearch/xpack/esql/analysis/VerifierTests.java:1123]: Add some negative testcases here for:

1. Cover more invalid input data types - all of the available DataTypes are defined in `DataType`. `AbstractFunctionTestCases` has the infrastructure to cover invalid input data types, it would be nice to have it, without it, `VerifierTests` is a good place to add them.
2. null as input
3. Invalid string literals as input

**Comment by fang-xing-esql** [comment]
I have a general question about `QSTR`, when I look through the testcases.  I wonder if we want to restrict the usage of `QSTR` only under a `WHERE` command/clause? Because all of the tests for `QSTR` seem use it only under `WHERE`. Internally a function expression can appear in a lot of places beyond `WHERE`. If we have this restriction, we might need more negative test cases to cover them. If we don't have this restriction, we might need more positive test cases to cover them too.

File: x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/function/fulltext/QueryStringFunction.java
Comment (fang-xing-esql) [reply x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/function/fulltext/QueryStringFunction.java:29]: > ++ Ioana on documentation and `AbstractFunctionTestCase` - we can probably defer this until we plan to move it out of snapshot builds

Can an issue be created to keep track of `AbstractFunctionTestCase` if it won't be included in this PR? It is important when this function is taken out of snapshot, we may not want to loose track of it. Thank you!

Comment (carlosdelest) [reply x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/function/fulltext/QueryStringFunction.java:29]: I've actually given it a go [here](https://github.com/elastic/elasticsearch/pull/112590/files#diff-bd42b7827384b735ff62baeb045ff9e37edcec201b5ee72f026e1cdb8c6496e8) in 4b97e6da93f1ad9d1e763b398ecb43befec45b30 - it performs type checking and folding.

File: x-pack/plugin/esql/qa/testFixtures/src/main/resources/qstr-function.csv-spec
Comment (carlosdelest) [reply x-pack/plugin/esql/qa/testFixtures/src/main/resources/qstr-function.csv-spec:2]: Understood! I did that in a3f859fa718ca48d0f6cbdddb436db468918d4d8

File: x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/analysis/Verifier.java
Comment (carlosdelest) [reply x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/analysis/Verifier.java]: Thanks for the pointer! Done in db1da96b92e06a16b780aa6d26df343149798c02

File: x-pack/plugin/esql/src/test/java/org/elasticsearch/xpack/esql/analysis/VerifierTests.java
Comment (carlosdelest) [reply x-pack/plugin/esql/src/test/java/org/elasticsearch/xpack/esql/analysis/VerifierTests.java]: Added tests in 97e1b8b5445bf5021405cd0216975da80e433938

**Comment by carlosdelest** [comment]
> I have a general question about `QSTR`, when I look through the testcases. I wonder if we want to restrict the usage of `QSTR` only under a `WHERE` command/clause? Because all of the tests for `QSTR` seem use it only under `WHERE`. Internally a function expression can appear in a lot of places beyond `WHERE`. If we have this restriction, we might need more negative test cases to cover them. If we don't have this restriction, we might need more positive test cases to cover them too.

You are correct, `QSTR` is going to be used just for `WHERE` commands. I added detection of this and the corresponding tests in c9d4741ce7e9c17f54ef72c8ed89f7c50e35c251

File: x-pack/plugin/esql/qa/testFixtures/src/main/resources/qstr-function.csv-spec
Comment (fang-xing-esql) [inline x-pack/plugin/esql/qa/testFixtures/src/main/resources/qstr-function.csv-spec:116]: A typo here? There is only one QSTR function in this query below.

File: x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/analysis/Verifier.java
Comment (fang-xing-esql) [inline x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/analysis/Verifier.java:685]: Just to be consistent with the other places in the code, use [{}] in the error message, the same applies to line 675.

File: x-pack/plugin/esql/src/internalClusterTest/java/org/elasticsearch/xpack/esql/plugin/QueryStringFunctionIT.java
Comment (fang-xing-esql) [inline x-pack/plugin/esql/src/internalClusterTest/java/org/elasticsearch/xpack/esql/plugin/QueryStringFunctionIT.java:31]: We don't validate the input string to `QSTR` before pushing it down to Lucene, I'd prefer we have some negative test cases for invalid input strings, so that we can confirm that users get meaningful error messages when they provide invalid query strings.

File: x-pack/plugin/esql/qa/testFixtures/src/main/resources/qstr-function.csv-spec
Comment (carlosdelest) [reply x-pack/plugin/esql/qa/testFixtures/src/main/resources/qstr-function.csv-spec:116]: I removed this tests as it was a copy/paste from the previous one - probably changed my mind mid-way and didn't remove it. Thanks for the catch!

File: x-pack/plugin/esql/src/test/java/org/elasticsearch/xpack/esql/expression/function/fulltext/QueryStringFunctionTests.java
Comment (carlosdelest) [reply x-pack/plugin/esql/src/test/java/org/elasticsearch/xpack/esql/expression/function/fulltext/QueryStringFunctionTests.java:28]: Question - do we need to generate docs for functions that are just used in snapshot builds? Is there a way to prevent these to end up in docs until we have them ready for our users?

File: x-pack/plugin/esql/src/internalClusterTest/java/org/elasticsearch/xpack/esql/plugin/QueryStringFunctionIT.java
Comment (carlosdelest) [reply x-pack/plugin/esql/src/internalClusterTest/java/org/elasticsearch/xpack/esql/plugin/QueryStringFunctionIT.java:32]: Makes sense, did that in e3427f00aea2470bf59489d32462e8b863d5501a

File: x-pack/plugin/esql/src/test/java/org/elasticsearch/xpack/esql/expression/function/fulltext/QueryStringFunctionTests.java
Comment (carlosdelest) [reply x-pack/plugin/esql/src/test/java/org/elasticsearch/xpack/esql/expression/function/fulltext/QueryStringFunctionTests.java:28]: I dug deeper on this - as long as the function is registered in the SnapshotRepository, no docs are generated for it. We need to make sure docs are generated when we remove QSTR from the snapshot build.

File: x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/analysis/Verifier.java
Comment (carlosdelest) [reply x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/analysis/Verifier.java]: Sure thing, thanks for the catch - 0c3e2f45c74873a7d92f25568199cbc745e96519

File: x-pack/plugin/esql/src/test/java/org/elasticsearch/xpack/esql/expression/function/fulltext/QueryStringFunctionTests.java
Comment (drewdaemon) [reply x-pack/plugin/esql/src/test/java/org/elasticsearch/xpack/esql/expression/function/fulltext/QueryStringFunctionTests.java:28]: We'll still want that documentation. As of https://github.com/elastic/elasticsearch/pull/112792 the plan was to provide the docs for snapshot functions, but to include the `preview` flag...

The reason is that we want the editor to be able to validate the function, even if it is in development. To do that, we need the function signature.

We will, however, not _suggest_ the function if `preview` is set to `true`.

Comment (carlosdelest) [reply x-pack/plugin/esql/src/test/java/org/elasticsearch/xpack/esql/expression/function/fulltext/QueryStringFunctionTests.java:28]: @drewdaemon @fang-xing-esql , currently snapshot functions do not generate documentation of any kind.

We don't want snapshot functions to be included in our ES docs, correct? IIUC, that would mean [changing the docs generation process](https://github.com/elastic/elasticsearch/blob/ef3a5a138530c43b8e6cd94585d13b853b819c57/x-pack/plugin/esql/src/test/java/org/elasticsearch/xpack/esql/expression/function/AbstractFunctionTestCase.java#L864) so Kibana docs are generated for all functions (including snapshot and preview). ES docs would keep being generated just for non-snapshot ones (including preview).

Is this correct? Is there work already in progress for doing this change, or should we do that on a separate PR?

**Review by ioanatia — APPROVED** [review]
looks good from my side - we do need to fix the conflicts before merging.
for the Kibana docs, let's start that conversation on whether we want all functions available in snapshot builds to have Kibana docs and then do any required changes separately.
I'd wait for @fang-xing-esql's approval since she invested more time than me reviewing this PR since it was opened.

File: x-pack/plugin/esql/src/test/java/org/elasticsearch/xpack/esql/expression/function/fulltext/QueryStringFunctionTests.java
Comment (fang-xing-esql) [reply x-pack/plugin/esql/src/test/java/org/elasticsearch/xpack/esql/expression/function/fulltext/QueryStringFunctionTests.java:28]: Snapshot functions are relatively new to us, thank you for putting a lot of effort on this @carlosdelest @ioanatia, it is in pretty good shape, and it will be a good example of snapshot functions moving forward. The only thing left that I can think of for now is docs.

There are a few steps below that we can do to make a docs available(if we decide to do it in this PR or as a follow up), if it can help Kibana to provide a better user experience, I'll rely on @drewdaemon 's input, as I have very little knowledge about how Kibana handles snapshot functions.

1. Add annotation to `QueryStringFunctionTests`, so that `AbstractFunctionTestCase` can recognize it an a registered function
```
@FunctionName("qstr")
public class QueryStringFunctionTests extends AbstractFunctionTestCase {
```

2. Make `AbstractFunctionTestCase` generate docs for snapshot functions.
```
    private static FunctionDefinition definition(String name) {
        EsqlFunctionRegistry registry = new EsqlFunctionRegistry().snapshotRegistry();
        if (registry.functionExists(name)) {
            return registry.resolveFunction(name);
        }
        return null;
    }
```

3. Include `ifeval::["{release-state}"=="unreleased"]` in the docs where this function will be linked in. There are a couple of examples for snapshot commands(inlinestats, lookup) here
```
ifeval::["{release-state}"=="unreleased"]
* experimental:[] <<esql-inlinestats-by>>
endif::[]
* <<esql-keep>>
* <<esql-limit>>
ifeval::["{release-state}"=="unreleased"]
* experimental:[] <<esql-lookup>>
endif::[]
```

ES|QL functions are in `esql-functions-operators.asciidoc`, `QSTR` can be linked under it, but don't forget to wrap is around with `ifeval::["{release-state}"=="unreleased"]` if it is still under snapshot.

**Review by fang-xing-esql — CHANGES_REQUESTED** [review]
Let's figure out the technical difficulty to have the docs in place in this PR, it is preferable to have it here so that the downstream components can use them more smoothly, but if it is too much effort we can separate them, just don't want to lose track of it.

File: x-pack/plugin/esql/src/test/java/org/elasticsearch/xpack/esql/expression/function/fulltext/QueryStringFunctionTests.java
Comment (carlosdelest) [reply x-pack/plugin/esql/src/test/java/org/elasticsearch/xpack/esql/expression/function/fulltext/QueryStringFunctionTests.java:30]: Thanks @fang-xing-esql ! I've generated the docs, and it seems that functions docs need to be explicitly linked from the [functions top level docs](https://github.com/elastic/elasticsearch/blob/b26d7d3e1489ffb35283603599a83cbb4762dcfe/docs/reference/esql/esql-functions-operators.asciidoc?plain=1#L86) - so we should be safe as the docs are there but not linked to the top-level pages.

I've merged main, resolved conflicts, and committed the generated docs that include Kibana docs.

LMK if this works! 👍

File: x-pack/plugin/esql/qa/testFixtures/src/main/resources/qstr-function.csv-spec
Comment (fang-xing-esql) [inline x-pack/plugin/esql/qa/testFixtures/src/main/resources/qstr-function.csv-spec:16]: A result tag(qstr-with-field-result) is missing, the tag is referenced in `docs/reference/esql/functions/examples/qstr.asciidoc`

File: x-pack/plugin/esql/src/test/java/org/elasticsearch/xpack/esql/expression/function/fulltext/QueryStringFunctionTests.java
Comment (fang-xing-esql) [inline x-pack/plugin/esql/src/test/java/org/elasticsearch/xpack/esql/expression/function/fulltext/QueryStringFunctionTests.java:29]: Thanks for including the generated docs.

Don't forget to commit the changes to `AbstractFunctionTestCase`, otherwise you may get these messages the next time `./gradlew -p x-pack/plugin/esql/ test` runs.

```
Changes not staged for commit:
  (use "git add/rm <file>..." to update what will be committed)
  (use "git restore <file>..." to discard changes in working directory)
	deleted:    docs/reference/esql/functions/description/qstr.asciidoc
	deleted:    docs/reference/esql/functions/examples/qstr.asciidoc
	deleted:    docs/reference/esql/functions/kibana/definition/qstr.json
	deleted:    docs/reference/esql/functions/kibana/docs/qstr.md
	deleted:    docs/reference/esql/functions/layout/qstr.asciidoc
	deleted:    docs/reference/esql/functions/parameters/qstr.asciidoc
	deleted:    docs/reference/esql/functions/signature/qstr.svg
	deleted:    docs/reference/esql/functions/types/qstr.asciidoc
```

**Comment by fang-xing-esql** [comment]
> > Let's figure out the technical difficulty to have the docs in place in this PR, it is preferable to have it here so that the downstream components can use them more smoothly, but if it is too much effort we can separate them, just don't want to lose track of it.
> 
> I think we have a strategy as the docs can be generated for Kibana but not included ES docs. LMKWYT!

Just added two more comments, we are almost there.

File: docs/reference/esql/functions/kibana/definition/qstr.json
Comment (fang-xing-esql) [inline docs/reference/esql/functions/kibana/definition/qstr.json:35]: Somehow the `"preview" : true` is not here, I'll update here if I find out why.

File: docs/reference/esql/functions/examples/qstr.asciidoc
Comment (fang-xing-esql) [inline docs/reference/esql/functions/examples/qstr.asciidoc:11]: This `qstr-with-field-result` tag is not in `str-function.csv-spec`.

File: docs/reference/esql/functions/layout/qstr.asciidoc
Comment (fang-xing-esql) [inline docs/reference/esql/functions/layout/qstr.asciidoc:7]: It is weird `VALUES` is mentioned here, should be `QSTR`.

File: x-pack/plugin/esql/src/test/java/org/elasticsearch/xpack/esql/expression/function/fulltext/QueryStringFunctionTests.java
Comment (drewdaemon) [reply x-pack/plugin/esql/src/test/java/org/elasticsearch/xpack/esql/expression/function/fulltext/QueryStringFunctionTests.java:30]: Looks great @carlosdelest thank you for your patience.

> We have two other functions available only in snapshot builds - rate and categorize - none of them have Kibana docs.

@ioanatia thank you for pointing this out!

This thread has highlighted that we confirm the user experience we want for snapshot functions in the editor and let that drive the documentation requirements. I will start that discussion outside of this PR.

Thanks all!

File: docs/reference/esql/functions/kibana/definition/qstr.json
Comment (carlosdelest) [reply docs/reference/esql/functions/kibana/definition/qstr.json:36]: This was due to the docs being generated without the merged version from main - fixed in 28b72f9e945ca27a824c6bb602a5e6e0e9f47bdf

File: docs/reference/esql/functions/layout/qstr.asciidoc
Comment (carlosdelest) [reply docs/reference/esql/functions/layout/qstr.asciidoc]: This was due to the docs being generated without the merged version from main - fixed in 28b72f9e945ca27a824c6bb602a5e6e0e9f47bdf

File: x-pack/plugin/esql/src/test/java/org/elasticsearch/xpack/esql/expression/function/fulltext/QueryStringFunctionTests.java
Comment (carlosdelest) [reply x-pack/plugin/esql/src/test/java/org/elasticsearch/xpack/esql/expression/function/fulltext/QueryStringFunctionTests.java:29]: I did that but reverted the change, as this implies changes from the Categorize function. I've opened [this PR](https://github.com/elastic/elasticsearch/pull/113080) for dealing with that specific change. Can you please review?

File: x-pack/plugin/esql/qa/testFixtures/src/main/resources/qstr-function.csv-spec
Comment (carlosdelest) [reply x-pack/plugin/esql/qa/testFixtures/src/main/resources/qstr-function.csv-spec:17]: I see, I added that in 70dd5d2a9610463857c4188d75c03a730c85f71c

File: x-pack/plugin/esql/src/test/java/org/elasticsearch/xpack/esql/expression/function/fulltext/QueryStringFunctionTests.java
Comment (fang-xing-esql) [reply x-pack/plugin/esql/src/test/java/org/elasticsearch/xpack/esql/expression/function/fulltext/QueryStringFunctionTests.java:29]: I guess we are pending on the PR to change `AbstractFunctionTestCase` to generate docs for snapshot functions, without it if just checking in `QSTR` related docs, it will be a surprise the next time people run `./gradlew -p x-pack/plugin/esql/ test` before pushing PRs, these docs will be marked as deleted.

Comment (carlosdelest) [reply x-pack/plugin/esql/src/test/java/org/elasticsearch/xpack/esql/expression/function/fulltext/QueryStringFunctionTests.java:29]: I just merged https://github.com/elastic/elasticsearch/pull/113080 👍

**Comment by carlosdelest** [comment]
@fang-xing-esql , I just merged https://github.com/elastic/elasticsearch/pull/113080 so docs generation can be done for `QSTR`.

Can you please check if there are any other pending concerns, or anything we should track for a follow-up PR, so we can merge this one?

Thanks!

**Review by fang-xing-esql — APPROVED** [review]
LGTM! Thank you for enabling full text search in ES|QL @carlosdelest !

**Comment by carlosdelest** [comment]
## 💚 All backports created successfully

| Status | Branch | Result |
|:------:|:------:|:------|
|✅|8.x|[](https://github.com/elastic/elasticsearch/pull/113189)|

### Questions ?
Please refer to the [Backport tool documentation](https://github.com/sorenlouv/backport)

**Pull request merged**
